satellite/metainfo/service.go

```diff
@@ -40,31 +41,121 @@ func (s *Service) Put(ctx context.Context, path string, pointer *pb.Pointer) (er
 
 	pointerBytes, err := proto.Marshal(pointer)
 	if err != nil {
-		return err
+		return Error.Wrap(err)
 	}
 
+	s.pointerDBWriteLock.Lock()
+	defer s.pointerDBWriteLock.Unlock()
+
 	// TODO(kaloyan): make sure that we know we are overwriting the pointer!
 	// In such case we should delete the pieces of the old segment if it was
 	// a remote one.
 	if err = s.DB.Put(ctx, []byte(path), pointerBytes); err != nil {
-		return err
+		return Error.Wrap(err)
 	}
 
 	return nil
 }
 
+// UpdatePieces atomically adds toAdd pieces and removes toRemove pieces from
+// the pointer under path. ref is the pointer that caller received via Get
+// prior to calling this method.
+//
+// It will first check if the pointer has been deleted or replaced. Then it
+// will remove the toRemove pieces and then it will add the toAdd pieces.
+// Replacing the node ID and the hash of a piece can be done by adding the
+// piece to both toAdd and toRemove.
+func (s *Service) UpdatePieces(ctx context.Context, path string, ref *pb.Pointer, toAdd, toRemove []*pb.RemotePiece) (pointer *pb.Pointer, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// lock the mutex to prevent Put and Delete to mess up with the pointer
+	// while it is updated here
+	s.pointerDBWriteLock.Lock()
```

Comment:
This won't be sufficient since we will break the satellite into separate servers.

Reply:
So what is the correct way to do it?

Reply:
There are a few ways to techincally implement them, but all of them require database support:

1. builtin transactions
2. compare-and-swap
3. compare-and-swap with versioning.

All of them roughly translate to code:

```
if !compare "old-value" and swap "new-value" {
    handle change conflict
}
```

Or in other words, you check whether the old-value has changed only update when it hasn't... otherwise, you need to handle the conflict.

Depending on the situation it might be that:
1. you load the value that is now in the database and redo your operation,
2. cancel the database change and clean-up (e.g. by deleting the newly uploaded piece)
3. reject the request

Reply:
So, we need to add a new method to the `KeyValueStore` interface:
```go
CompareAndSwap(ctx context.Context, key Key, oldValue, newValue Value) error
```
and implement it correctly in all database implementations.

Is this correct?

Reply:
Yeah

Reply:
How would this affect `Put` and `Delete`?

I guess we'd want to `Put` only if there is no such key yet -- to prevent overwriting changes that were just added by another process/goroutune. Perhaps, `Put` should be implemented as `CompareAndSwap(ctx, key, nil, value)`.

How about `Delete`? Should we be worried about anything there?

Reply:
Yeah, put should fail if something already exists, so yes.

For the delete yeah, it's somewhat equivalent to `CompareAndSwap(ctx, key, oldvalue, nil)`. But, sometimes nil value has meaning. I don't think there's much more to delete.

Reply:
I created https://github.com/storj/storj/pull/2602 for adding `CompareAndSwap`.

Reply:
This now uses the new `CompareAndSwap`.